src/software/sensor_fusion/filter/ball_filter.cpp

```diff
@@ -33,6 +34,42 @@ void BallFilter::addNewDetectionsToBuffer(std::vector<BallDetection> new_ball_de
     // recent detections (largest timestamp) at the end of the buffer.
     std::sort(new_ball_detections.begin(), new_ball_detections.end());
 
+    // reject a single high residual detection if we have multiple detections and
+    // enough information to reliably reject it
+    // do-while loop so we can early out when we need to
+    do
+    {
+        if (ball_detection_buffer.size() >= MIN_BUFFER_SIZE &&
+            new_ball_detections.size() > 1)
+        {
+            boost::circular_buffer<BallDetection> all_detections(
+                ball_detection_buffer.size() + new_ball_detections.size(),
+                BallDetection());
+            auto next_it = std::copy(ball_detection_buffer.begin(),
+                                     ball_detection_buffer.end(), all_detections.begin());
+            std::copy(new_ball_detections.begin(), new_ball_detections.end(), next_it);
+            auto regression_results = calculateLinearRegression(all_detections);
+            auto velocity_estimate =
+                estimateBallVelocity(all_detections, regression_results.regression_line);
+            if (velocity_estimate && velocity_estimate->average_velocity.length() > 1.0)
+            {
+                break;
+            }
+            // find the highest residual detection out of the new ones
+            auto highest_residual_it =
+                std::max_element(new_ball_detections.begin(), new_ball_detections.end(),
+                                 [&regression_results](const auto &lhs_detection,
+                                                       const auto &rhs_detection) {
+                                     return distance(lhs_detection.position,
+                                                     regression_results.regression_line) <
+                                            distance(rhs_detection.position,
+                                                     regression_results.regression_line);
+                                 });
+            // erase it
+            new_ball_detections.erase(highest_residual_it);
+        }
+    } while (false);
```

Comment:
seems like this never loops?

Reply:
this was there so that i could early exit out of that section of code, tunrs out it was unncessary